### PR TITLE
Make tests run faster?

### PR DESCRIPTION
**Pull request by mehulkar**
- Use `turbo` to run unit, integration, e2e tests so we can _try_ to get Full Turbo in some places
- at the very least, stop building `turbo` 7 extra times times

    we build turbo just to _kick off_ tests, and then each of the test suites also builds turbo for each matrix run. (2 for unit, 3 for unit and e2e each, and 1 for integration)

### CODE REVIEW

File: .github/workflows/test.yml
Comment (NicholasLYang) [inline .github/workflows/test.yml:235]: This won't matter soon (hopefully)

File: cli/turbo.json
Comment (NicholasLYang) [inline cli/turbo.json:45]: Any reason we're specifying order instead of letting them run concurrently?

**Review by NicholasLYang — APPROVED** [review]
This makes sense to me. At some point I think it might be nice to have all of our tests running via Turbo in parallel, but moving unit, integration and e2e is a good start.

File: cli/turbo.json
Comment (mehulkar) [reply cli/turbo.json:45]: The logs are a bit harder to parse (before `--log-order=group` lands) and I wasn't 100% sure that it was safe to run in parallel, so i wanted to start in easy mode first.

File: cli/Makefile
Comment (chris-olszewski) [inline cli/Makefile:151]: This is the same as `e2e-prebuilt` now

File: cli/Makefile
Comment (chris-olszewski) [inline cli/Makefile:316]: Just calling out that this target still depends on `turbo`

File: .github/workflows/test.yml
Comment (chris-olszewski) [inline .github/workflows/test.yml:216]: I think we still want to restrict this to run only when either turborepo (Rust or Go) changes

File: cli/turbo.json
Comment (chris-olszewski) [inline cli/turbo.json:6]: Not necessarily related to these changes, but in the same ballpark. Should we change our build script? Right now we're just running `make` which produces a ton of junk that's not the turbo executable. If we're being honest we should also list out all of the Rust crates that end up getting used when building our Rust and Go binary

File: cli/turbo.json
Comment (chris-olszewski) [inline cli/turbo.json:23]: Reading into this, we need this ordering since `integration-tests` will trigger a turbo binary build which is needed for the `e2e` tests, right?

**Review by chris-olszewski — COMMENTED** [review]
It feels a little weird duplicating the source files as inputs for the tests. Would it be cleaner to try to properly setup `turbo#build` to track all of the changes that can produce a different turbo binary and then have our tests depend on `build`

File: cli/Makefile
Comment (mehulkar) [reply cli/Makefile:316]: yeah i don't know why it's duplicated here

File: .github/workflows/test.yml
Comment (mehulkar) [reply .github/workflows/test.yml:216]: i was thinking turborepo can just do this. "always run, but get cache hit" when source files haven't changed

File: cli/turbo.json
Comment (mehulkar) [reply cli/turbo.json:23]: ideally we don't actually need the ordering, since `build` happens before tests are run entirely. the ordering is ONLY for DX right now

File: cli/turbo.json
Comment (mehulkar) [reply cli/turbo.json]: added all the crates and stuff (need to verify that it actually invalidates cache on the crates changing)

**Comment by mehulkar** [comment]
> It feels a little weird duplicating the source files as inputs for the tests. Would it be cleaner to try to properly setup turbo#build to track all of the changes that can produce a different turbo binary and then have our tests depend on build

That's a good idea, although actually running `cli#build` is a catch 22, since our entry point is a local build of turbo (via `pnpm -- turbo`, so it's already built

**Comment by mehulkar** [comment]
@chris-olszewski I just opted out of `pnpm -- turbo` and used `pnpm run test` instead, and setup the `test` => `build` dependency as you suggested. I _think_ this should work, but not sure if it impacts other test suites

File: .github/workflows/test.yml
Comment (mehulkar) [reply .github/workflows/test.yml:227]: how soon? Should I find a way to run both in this PR or block this PR until we don't need it anymore? Is https://github.com/vercel/turbo/pull/4181 the thing that makes it so we don't need both?

Comment (NicholasLYang) [reply .github/workflows/test.yml:227]: Yeah once we flip the rust switch (hopefully) we won't switch it back and we can start deleting Go code. I'd wait until the switch PR lands and then we can merge this

File: cli/turbo.json
Comment (chris-olszewski) [reply cli/turbo.json]: Ah, I forgot that `pnpm -- turbo run test` will end up building turbo right away.

File: .github/workflows/test.yml
Comment (chris-olszewski) [reply .github/workflows/test.yml:207]: My only concern with that is that we'll be waiting on a turbo build before we can decide if we should run.

**Pull request closed without merging**
